### PR TITLE
Allow per-IP TCP sockets

### PR DESCRIPTION
**Pull request by rkday**
Adds a '-t ti' option that's the TCP analogue of '-t ui'.

### CODE REVIEW

File: src/call.cpp [hunk 1]
```diff
@@ -941,6 +941,38 @@ bool call::connect_socket_if_needed()
 
         sipp_customize_socket(call_socket);
 
+        if (peripsocket) {
+            struct sockaddr_storage saddr;
+            char peripaddr[256];
```
Comment (vodik) [inline src/call.cpp]: Shouldn't this be `SIPP_MAX_MSG_SIZE`? Or am I mistaken? To my understanding, it looks like `getFieldFromInputFile` assumes that buffer length.

File: src/call.cpp [hunk 2]
```diff
@@ -941,6 +941,38 @@ bool call::connect_socket_if_needed()
...
+            char *tmp = peripaddr;
+            getFieldFromInputFile(ip_file, peripfield, NULL, tmp);
+            struct addrinfo * h ;
+            struct addrinfo   hints;
+            memset((char*)&hints, 0, sizeof(hints));
+            hints.ai_flags  = AI_PASSIVE;
+            hints.ai_family = PF_UNSPEC;
```
Comment (vodik) [inline src/call.cpp]: This should be `AF_UNSPEC`. The distinction is kinda moot but we're talking about the address family, not the protocol family.

**Comment by vodik** [comment]
Have to say, looks useful.

File: src/call.cpp [hunk 3]
```diff
@@ -941,6 +941,38 @@ bool call::connect_socket_if_needed()
...
         sipp_customize_socket(call_socket);
 
+        if (peripsocket) {
+            struct sockaddr_storage saddr;
+            char peripaddr[256];
+            char *tmp = peripaddr;
+            getFieldFromInputFile(ip_file, peripfield, NULL, tmp);
+            struct addrinfo * h ;
+            struct addrinfo   hints;
+            memset((char*)&hints, 0, sizeof(hints));
```
Comment (wdoekes) [inline src/call.cpp]: Use struct initializer `... hints = {0,}` and drop the memset.

File: src/call.cpp [hunk 4]
```diff
@@ -941,6 +941,38 @@ bool call::connect_socket_if_needed()
 
         sipp_customize_socket(call_socket);
 
+        if (peripsocket) {
+            struct sockaddr_storage saddr;
+            char peripaddr[256];
+            char *tmp = peripaddr;
+            getFieldFromInputFile(ip_file, peripfield, NULL, tmp);
```
Comment (wdoekes) [inline src/call.cpp]: What good does the addition of `char *tmp = ...` do? Looks to me like `peripaddr` could simply be passed twice as args 2 and 4..

(And I believe we settled on `char*`.)

File: src/call.cpp [hunk 5]
```diff
@@ -941,6 +941,38 @@ bool call::connect_socket_if_needed()
...
+            struct addrinfo * h ;
+            struct addrinfo   hints;
+            memset((char*)&hints, 0, sizeof(hints));
+            hints.ai_flags  = AI_PASSIVE;
+            hints.ai_family = PF_UNSPEC;
+            getaddrinfo(peripaddr,
+                        NULL,
+                        &hints,
+                        &h);
```
Comment (wdoekes) [inline src/call.cpp]: A bit superfluous indent, imo.

File: src/call.cpp [hunk 6]
```diff
@@ -941,6 +941,38 @@ bool call::connect_socket_if_needed()
...
+            if (use_ipv6) {
+                (_RCAST(struct sockaddr_in6 *, &saddr))->sin6_port = htons(0);
+            } else {
+                (_RCAST(struct sockaddr_in *, &saddr))->sin_port = htons(0);
+            }
+            int port = -1;
+            int ret = sipp_bind_socket(call_socket, &saddr, &port);
+             if (ret != 0) {
+              WARNING_NO("Could not bind to %s: %d", peripaddr, ret);
+            }
```
Comment (wdoekes) [inline src/call.cpp]: Bad indentation.

File: src/call.cpp [hunk 7]
```diff
@@ -941,6 +941,38 @@ bool call::connect_socket_if_needed()
...
+        }
+
+
```
Comment (wdoekes) [inline src/call.cpp]: I'd favor it if the whole getaddrinfo thing got moved out of here. I believe I've seen the same getaddrinfo code in various places.

```
if (peripsocket) {
    struct sockaddr_storage saddr;
    char peripaddr[256];
    getFieldFromInputFile(ip_file, peripfield, NULL, peripfield);
    sipp_aton(use_ipv6, &saddr, peripfield);
    int port = -1;
    int ret = sipp_bind_socket(call_socket, &saddr, &port);
    ...
```

File: src/sipp.cpp [hunk 8]
```diff
@@ -1567,9 +1567,6 @@ int main(int argc, char *argv[])
                     break;
                 }
 
-                if (peripsocket && transport != T_UDP) {
-                    ERROR("You can only use a perip socket with UDP!\n");
-                }
```
Comment (wdoekes) [inline src/sipp.cpp]: Does it work for all transports? Or just for T_UDP and T_TCP ?

File: src/call.cpp (hunk 2 above)
Comment (wdoekes) [inline src/call.cpp]: For the record:

socket(2) says PF_* is deprecated, so AF_* should be preferred anyway.
In line with those docs, e.g. packet(7) uses only AF_PACKET in the manual.

File: src/call.cpp (hunk 7 above)
Comment (vodik) [inline src/call.cpp]: Yeah, talking about refactoring, thus could use some work too. For example, we're passing an empty service field to `getaddrinfo` and then doing AF-dependant magic to inject the desired port.

File: src/call.cpp (hunk 1 above)
Comment (rkday) [inline src/call.cpp]: I've changed getFieldFromInputFile to take a length parameter - we don't need 64kb for an IP address.

File: src/call.cpp (hunk 4 above)
Comment (rkday) [inline src/call.cpp]: It couldn't previously (getFieldFromInputFile took a reference to a char*), but I've refactored to make it better.

File: src/call.cpp (hunk 7 above)
Comment (rkday) [inline src/call.cpp]: I've pulled this logic ("take an IP address char* and fill in a struct sockaddr_storage") out into a separate function, although I haven't used it everywhere yet.

File: src/call.cpp (hunk 1 above)
Comment (vodik) [inline src/call.cpp]: Yeah, even better.

File: src/call.cpp [hunk 9]
```diff
@@ -904,27 +895,13 @@ bool call::connect_socket_if_needed()
...
-                       _RCAST(struct sockaddr_storage *,h->ai_addr)));
-
-            if (use_ipv6) {
-                (_RCAST(struct sockaddr_in6 *, &saddr))->sin6_port = htons(local_port);
-            } else {
-                (_RCAST(struct sockaddr_in *, &saddr))->sin_port = htons(local_port);
-            }
+                    local_addr_storage->ai_addr,
+                    SOCK_ADDR_SIZE(
+                        _RCAST(struct sockaddr_storage *,local_addr_storage->ai_addr)));
```
Comment (vodik) [inline src/call.cpp]:
~~~c
memcpy(&saddr, local_addr_storage->ai_addr, sizeof(struct sockaddr_storage));
~~~

Is perfectly safe and simpler.

File: src/socket.cpp [hunk 10]
```diff
@@ -510,13 +510,9 @@ void setup_ctrl_socket()
 
     memset(&ctl_sa, 0, sizeof(struct sockaddr_storage));
     if (control_ip[0]) {
-        struct addrinfo hints;
+        struct addrinfo hints = {AI_PASSIVE, AF_UNSPEC,};
```
Comment (vodik) [inline src/socket.cpp]: I really wish we had C style struct initialization in C++... :/

File: src/socket.cpp [hunk 11]
```diff
@@ -3076,6 +2963,29 @@ bool sipMsgCheck (const char *P_msg, struct sipp_socket *socket)
     return false;
 }
 
+bool fill_sockaddr_from_ip(struct sockaddr_storage* saddr, char* ip, bool use_ipv6)
+{
+    struct addrinfo* h ;
+    struct addrinfo hints = {AI_PASSIVE, AF_UNSPEC,};
+    if (!getaddrinfo(ip, NULL, &hints, &h))
```
Comment (vodik) [inline src/socket.cpp]:
~~~c
char service[std::numeric_limits<unsigned short>::digits10 + 1];
snprintf(service, sizeof(service), "%hs", local_port);
if (!getaddrinfo(ip, service, &hints, &h);
//...
~~~

And we don't have to worry about any of the problems now with `(_RCAST(struct sockaddr_in6 *, saddr))`. That said, I'd also consider this a temporary fix, ideally we should keep the port as a number.

**Comment by wdoekes** [comment]
(And don't forget to squash before merging, since the commits aren't logical changesets.)

File: src/call.cpp [hunk 12]
```diff
@@ -855,7 +855,7 @@ bool call::connect_socket_if_needed()
     if(!multisocket) return true;
 
     if(transport == T_UDP) {
-        struct sockaddr_storage saddr;
+        struct sockaddr_storage saddr = {0,};
```
Comment (vodik) [inline src/call.cpp]:
~~~c
struct sockaddr_storage saddr = {};
~~~

Is enough in c++11.

File: src/call.cpp [hunk 13]
```diff
@@ -881,50 +879,29 @@ bool call::connect_socket_if_needed()
...
+                    SOCK_ADDR_SIZE(
+                        _RCAST(struct sockaddr_storage *,local_addr_storage->ai_addr)));
+        
             if (use_ipv6) {
-                (_RCAST(struct sockaddr_in6 *, &saddr))->sin6_port = htons(local_port);
+                saddr.ss_family       = AF_INET6;
             } else {
-                (_RCAST(struct sockaddr_in *, &saddr))->sin_port = htons(local_port);
+                saddr.ss_family       = AF_INET;
             }
```
Comment (vodik) [inline src/call.cpp]: The setting of `ss_family` shouldn't be necessary, it should always be set.

File: src/call.cpp [hunk 14]
```diff
@@ -881,50 +879,29 @@ bool call::connect_socket_if_needed()
...
+                    local_addr_storage->ai_addr,
+                    SOCK_ADDR_SIZE(
+                        _RCAST(struct sockaddr_storage *,local_addr_storage->ai_addr)));
+        
             if (use_ipv6) {
-                (_RCAST(struct sockaddr_in6 *, &saddr))->sin6_port = htons(local_port);
+                saddr.ss_family       = AF_INET6;
             } else {
-                (_RCAST(struct sockaddr_in *, &saddr))->sin_port = htons(local_port);
+                saddr.ss_family       = AF_INET;
```
Comment (vodik) [inline src/call.cpp]: The setting of `ss_family` shouldn't be necessary, it should always be set.

File: src/call.cpp [hunk 15]
```diff
@@ -881,50 +879,29 @@ bool call::connect_socket_if_needed()
...
+            fill_sockaddr_from_ip(&saddr, peripaddr, use_ipv6 ? AF_INET6 : AF_INET);
+        } else {
             memcpy(&saddr,
-                   h->ai_addr,
-                   SOCK_ADDR_SIZE(
-                       _RCAST(struct sockaddr_storage *,h->ai_addr)));
-
+                    local_addr_storage->ai_addr,
+                    SOCK_ADDR_SIZE(
+                        _RCAST(struct sockaddr_storage *,local_addr_storage->ai_addr)));
```
Comment (vodik) [inline src/call.cpp]: #149 kills the `SOCK_ADDR_SIZE` macro, and we don't need it here anyways:

```c
memcpy(&saddr, local_addr_storage->ai_addr, local_addr_storage->ai_addrlen);

File: src/socket.cpp [hunk 16]
```diff
@@ -2609,111 +2567,44 @@ int open_connections()
...
+
+        // Bind socket to local_ip
+        if (peripsocket) {
+            // On some machines it fails to bind to the self computed local
+            // IP address.
+            // For the socket per IP mode, bind the main socket to the
+            // first IP address specified in the inject file.
+            inFiles[ip_file]->getField(0, peripfield, peripaddr, sizeof(peripaddr));
+            if (!fill_sockaddr_from_ip(&local_sockaddr, peripaddr, local_ip_is_ipv6 ? AF_INET6 : AF_INET))
+            {
```
Comment (vodik) [inline src/socket.cpp]: Maybe we want to have a consistent style on brackets?

File: src/socket.cpp [hunk 17]
```diff
@@ -3076,6 +2905,29 @@ bool sipMsgCheck (const char *P_msg, struct sipp_socket *socket)
...
+    if (h->ai_family == AF_INET6) {
+        (_RCAST(struct sockaddr_in6 *, saddr))->sin6_port = htons(local_port);
+        saddr->ss_family       = AF_INET6;
+    } else {
+        (_RCAST(struct sockaddr_in *, saddr))->sin_port = htons(local_port);
+        saddr->ss_family       = AF_INET;
+    }
+    freeaddrinfo(h);
+    return true;
+}
```
Comment (vodik) [inline src/socket.cpp]: Yeah, if we're copying `ai_addr` into `saddr`, `ss_family` will be set already.

**Comment by vodik** [comment]
Very nice, this really cleans up the code base.

**Comment by wdoekes** [comment]
If this is still useful, it will require a rebase and a squash.

**Comment by wdoekes** [comment]
The merge conflicts are overwhelming. Assuming this is left for dead. Feel free to reopen :)

**Pull request closed without merging**